internal/logging/ring_buffer.go

```diff
@@ -0,0 +1,52 @@
+package logging
+
+// Ring buffer structure, but without a start position for reading. All reads are tail reads with
+// no concept for what has already been read. The use case here is for logging to keep track of the
+// last X bytes written.
+type ringBuffer struct {
+	buffer []byte
+	size   int
+	end    int
+	full   bool
+}
+
+func newRingBuffer(size int) *ringBuffer {
+	return &ringBuffer{make([]byte, size), size, 0, false}
+}
+
+// Read reads the last len(p) bytes from the buffer.
+// Typically len(p) should be the size of the buffer.
+// This method satisfies the io.Reader interface.
+func (b *ringBuffer) Read(p []byte) (int, error) {
```

Comment:
This feels like it's overkill for what we need. Is this just so that we can satisfy io.Reader? What is satisfying that interface giving us?

Reply:
What do you mean by "this" in "Is this just so ..."?

The contents of this method are required for reading correctly from the ring buffer. The signature is simply to satisfy the io.Reader interface because I thought that would be a good idea.

Reply:
Fair enough.